op-deployer/pkg/deployer/upgrade/v200/upgrade.go

```diff
@@ -0,0 +1,61 @@
+package v200
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/ethereum-optimism/optimism/op-chain-ops/script"
+	"github.com/ethereum-optimism/optimism/op-deployer/pkg/deployer/opcm"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/lmittmann/w3"
+)
+
+type UpgradeOPChainInput struct {
+	Prank               common.Address  `json:"prank"`
+	Opcm                common.Address  `json:"opcm"`
+	EncodedChainConfigs []OPChainConfig `evm:"-" json:"chainConfigs"`
+}
+
+type OPChainConfig struct {
+	SystemConfigProxy common.Address `json:"systemConfigProxy"`
+	ProxyAdmin        common.Address `json:"proxyAdmin"`
+	AbsolutePrestate  common.Hash    `json:"absolutePrestate"`
+}
+
+var opChainConfigEncoder = w3.MustNewFunc("dummy((address systemConfigProxy,address proxyAdmin,bytes32 absolutePrestate)[])", "")
+
+func (u *UpgradeOPChainInput) OpChainConfigs() ([]byte, error) {
+	data, err := opChainConfigEncoder.EncodeArgs(u.EncodedChainConfigs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encode chain configs: %w", err)
+	}
+	return data[4:], nil
+}
```

Comment:
Can you help me understand the use of "dummy" function here? The name makes it seem like it belongs in a test file but is it being used in the production code?

Reply:
Related question not specific to this pr but more for the op-deployer design as a whole: why do we invoke a solidity script from go to execute the tx instead of just sending the tx directly with the go-ethereum or w3 package? The solidity script seems much more heavyweight, but I know there must be some advantages that I’m not thinking of that make it a good trade off

Reply:
The `w3` lib only works with functions, not structs, so we need to provide a function name to `MustNewFunc` for it to parse the input correctly. The selector gets trimmed off via `data[4:]`, thus leaving behind just the calldata for the arguments.

Reply:
Doesn't go-ethereum allow you to encode structs as calldata though? Whats the advantage of using solidity scripts to send the tx? To me it seems like added complexity through additional layers of indirection compared to what I'm used to. I'm wondering if its just a personal preference to use solidity scripts of if there are some clear advantages

Reply:
go-ethereum does allow you to do this, but the API is very complex since it assumes you're using codegen.

Using solidity scripts is better because it allows us to simulate the transaction prior to sending it. It also makes it easier to perform pre-flight checks, since the scripts have access to Ethereum state without having to manually craft eth_calls in Golang.